Prometheus/TextSerializer.cs

```diff
@@ -49,19 +56,77 @@ public async Task WriteFamilyDeclarationAsync(byte[][] headerLines, Cancellation
         // Size limit guided by https://stackoverflow.com/questions/21146544/what-is-the-maximum-length-of-double-tostringd
         private readonly byte[] _stringBytesBuffer = new byte[32];
 
-        // name{labelkey1="labelvalue1",labelkey2="labelvalue2"} 123.456
-        public async Task WriteMetricAsync(byte[] identifier, double value, CancellationToken cancel)
+        // 123.456
+        // Note: Terminates with a NEWLINE
+        public async Task WriteValuePartAsync(double value, CancellationToken cancel)
         {
-            await _stream.Value.WriteAsync(identifier, 0, identifier.Length, cancel);
-            await _stream.Value.WriteAsync(Space, 0, Space.Length, cancel);
-
             var valueAsString = value.ToString(CultureInfo.InvariantCulture);
-
             var numBytes = PrometheusConstants.ExportEncoding
                 .GetBytes(valueAsString, 0, valueAsString.Length, _stringBytesBuffer, 0);
 
             await _stream.Value.WriteAsync(_stringBytesBuffer, 0, numBytes, cancel);
             await _stream.Value.WriteAsync(NewLine, 0, NewLine.Length, cancel);
         }
+
+        /// <summary>
+        /// Creates a metric identifier, with an optional name postfix and an optional extra label to append to the end.
+        /// familyname_postfix{labelkey1="labelvalue1",labelkey2="labelvalue2"}
+        /// Note: Terminates with a SPACE
+        /// </summary>
+        public async Task WriteIdentifierPartAsync(byte[] name, byte[] flatennedLabels, CancellationToken cancel,
+            byte[]? postfix = null, byte[]? extraLabelName = null, byte[]? extraLabelValue = null,
+            byte[]? extraLabelValueOpenMetrics = null)
```

Comment:
extraLabelValue for Prometheus and extraLabelValueOpenMetrics for OpenMetrics is the pattern, right? If so, I would prefer "extraLabelValuePrometheus" for the former, to treat the two equally.

Alternatively, can we provide some DTO here that contains both values in one argument? This special logic for the extra label here might be easier to understand that way.

Alternatively, can the caller already know which output format we use and only provide the right value?

Reply:
Added DTO 👍